File: src/player.rs
//! 时序调度：把曲谱展开成带绝对时间戳的动作表，再按表执行。
//!
//! 设计要点：
//! 1. 时间戳全部相对演奏起点计算（绝对时间轴），不累加 sleep 误差；
//! 2. 相邻同修饰的音符合并修饰键区间，减少鼠标事件；
//! 3. 完全按谱面时值发按键，不做顺延或延长；
//! 4. 任何退出路径都释放按键。

use std::sync::{Arc, Mutex};
use std::thread;
use std::time::{Duration, Instant};

use crate::config::{Config, MouseButton, key_name};
use crate::input;
use crate::score::{Event, Pitch, Register, Song};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Action {
    KeyDown(u16),
    KeyUp(u16),
    MouseDown(MouseButton),
    MouseUp(MouseButton),
}

#[derive(Debug, Clone, Copy)]
pub struct TimedAction {
    pub at_ms: f64,
    pub action: Action,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PlayState {
    Running,
    Stopped,
}

pub type Control = Arc<Mutex<PlayState>>;

struct Note {
    start_ms: f64,
    dur_ms: f64,
    vk: u16,
    buttons: Vec<MouseButton>,
}

/// 音区整体平移（半音阶十二度之外的粗移调）：-1 低八度，+1 高八度
fn shift_register(reg: Register, shift: i32) -> Register {
    let idx = match reg {
        Register::Low => 0,
        Register::Mid => 1,
        Register::High => 2,
        Register::Top => 3,
    } + shift;
    match idx.clamp(0, 3) {
        0 => Register::Low,
        1 => Register::Mid,
        2 => Register::High,
        _ => Register::Top,
    }
}

/// 按键时长的数学下限（毫秒）。时值比 `note_gap_ms` 还短的音会被压到这里，
/// 保证「按下」之后一定有对应的「松开」—— 否则键会卡在按下状态。
/// 这不是保底，是防止动作表出现「先松后按」的硬约束。
const MIN_KEY_HOLD_MS: f64 = 1.0;

/// 把曲谱展开成动作表。
///
/// 完全按谱面时值来：不缩短、不顺延、不延长。只有两类顺序约束会推迟动作 ——
/// 修饰键的切换必须等前一个音彻底松开，段首音符必须等修饰键建立起来，
/// 否则会发出滑音或整段跑调。这两条不改变总时长，只决定先后。
pub fn build_actions(song: &Song, cfg: &Config, shift: i32) -> Vec<TimedAction> {
    let ms_per_tick = song.ms_per_tick();
    let mut notes: Vec<Note> = Vec::new();
    // 时间轴整体后移一个 lead，保证第一个音的鼠标修饰键也能提前按下
    let mut cursor = cfg.timing.mouse_lead_ms as f64;

    for ev in &song.events {
        let dur = ev.ticks() as f64 * ms_per_tick;
        if let Event::Note { pitch, .. } = ev {
            let p = Pitch {
                register: shift_register(pitch.register, shift),
                degree: pitch.degree,
                sharp: pitch.sharp,
            };
            let (vk, buttons) = cfg.resolve(p);
            notes.push(Note {
                start_ms: cursor,
                dur_ms: dur,
                vk,
                buttons,
            });
        }
        cursor += dur;
    }

    let lead = cfg.timing.mouse_lead_ms as f64;
    let gap = cfg.timing.note_gap_ms as f64;

    let mut actions: Vec<TimedAction> = Vec::new();
    // 修饰键按"差集"维护：本段仍然需要的键保持按住不撒手，
    // 只松开不再需要的、按下新需要的。
    let mut held: Vec<MouseButton> = Vec::new();
    let mut last_off = 0.0_f64;
    // 上一段最后一个音符键的松开时刻。这一段修饰键要动，必须等它过去。
    let mut prev_last_key_up = 0.0_f64;
    let mut i = 0;

    while i < notes.len() {
        let btns = notes[i].buttons.clone();
        let mut j = i;
        while j + 1 < notes.len() && notes[j + 1].buttons == btns {
            j += 1;
        }

        // 修饰键最早能动的时刻：前一个音彻底结束，再留 gap 的余量。
        // 早了的话，音符还在响、音高就变了，会发出一个很难听的滑音。
        let earliest = if i > 0 { prev_last_key_up + gap } else { 0.0 };
        let on_at = (notes[i].start_ms - lead).max(earliest);

        // 先松开本段用不到的，再按下本段新增的。
        // 两者同刻时靠 order() 保证「先按新的、再松旧的」，不会瞬间串音。
        for b in held.iter().copied() {
            if !btns.contains(&b) {
                actions.push(TimedAction {
                    at_ms: on_at,
                    action: Action::MouseUp(b),
                });
            }
        }
        for b in btns.iter().copied() {
            if !held.contains(&b) {
                actions.push(TimedAction {
                    at_ms: on_at,
                    action: Action::MouseDown(b),
                });
            }
        }
        held = btns.clone();

        // 音符键。段首那一个必须等修饰键建立起来（on_at + lead），
        // 否则开头一小段是按在错误的音区上，听起来就是跑调。
        let mut last_key_up = 0.0_f64;
        for idx in i..=j {
            let n = &notes[idx];
            let down_at = if idx == i {
                n.start_ms.max(on_at + lead)
            } else {
                n.start_ms
            };
            // 完全按谱面时值：按住 = 时值 - 松键间隔
            let up_at = down_at + (n.dur_ms - gap).max(MIN_KEY_HOLD_MS);

            actions.push(TimedAction {
                at_ms: down_at,
                action: Action::KeyDown(n.vk),
            });
            actions.push(TimedAction {
                at_ms: up_at,
                action: Action::KeyUp(n.vk),
            });
            last_key_up = up_at;
        }
        prev_last_key_up = last_key_up;

        // 收尾松开修饰键，同样要等本段最后一个音真的松开之后
        last_off = (notes[j].start_ms + notes[j].dur_ms + lead).max(last_key_up + gap);
        i = j + 1;
    }

    // 收尾：松开还按着的修饰键
    for b in held {
        actions.push(TimedAction {
            at_ms: last_off,
            action: Action::MouseUp(b),
        });
    }

    actions.sort_by(|a, b| {
        a.at_ms
            .total_cmp(&b.at_ms)
            .then_with(|| order(a.action).cmp(&order(b.action)))
    });
    actions
}

/// 同一时刻的执行顺序。
/// 松开音符键排最前：先掐断上一个音，再建立下一个音的修饰，避免瞬间串音。
/// 松开修饰键排最后：保证修饰在音符键松开之后才释放。
fn order(a: Action) -> u8 {
    match a {
        Action::KeyUp(_) => 0,
        Action::MouseDown(_) => 1,
        Action::KeyDown(_) => 2,
        Action::MouseUp(_) => 3,
    }
}

/// 按动作表演奏。dry_run 时只打印不发送输入。
/// progress 用于把当前演奏位置回传给界面（毫秒）。
pub fn play(
    actions: &[TimedAction],
    cfg: &Config,
    control: Control,
    dry_run: bool,
    progress: Option<&std::sync::atomic::AtomicU64>,
) -> Result<(), String> {
    // 无论正常结束、提前 return 还是 panic，离开作用域时都会松开所有按键。
    // 漏掉这一步的后果很严重：鼠标键会卡在按下状态，导致整个系统点不动东西。
    let keys = cfg.all_keys();
    let _guard = ReleaseGuard {
        keys: &keys,
        active: !dry_run,
    };

    let base = Instant::now();
    let mut idx = 0_usize;

    loop {
        if *control.lock().map_err(|_| "控制状态锁已损坏")? == PlayState::Stopped {
            break;
        }

        if idx >= actions.len() {
            break;
        }

        let target = base + Duration::from_secs_f64(actions[idx].at_ms / 1000.0);
        let now = Instant::now();
        if now < target {
            let remain = target - now;
            thread::sleep(remain.min(Duration::from_millis(20)));
            continue;
        }

        let ta = &actions[idx];
        if dry_run {
            println!("  [{:8.0} ms] {}", ta.at_ms, describe(ta.action));
        } else {
            dispatch(ta.action);
        }
        if let Some(p) = progress {
            p.store(ta.at_ms as u64, std::sync::atomic::Ordering::Relaxed);
        }
        idx += 1;
    }

    Ok(())
}

/// 作用域结束时松开所有按键，panic 展开时同样生效。
pub struct ReleaseGuard<'a> {
    keys: &'a [u16],
    /// dry-run 不发送任何输入，这里也要跟着关掉
    active: bool,
}

impl<'a> ReleaseGuard<'a> {
    pub fn new(keys: &'a [u16], active: bool) -> Self {
        ReleaseGuard { keys, active }
    }
}

impl Drop for ReleaseGuard<'_> {
    fn drop(&mut self) {
        if self.active {
            input::release_all(self.keys);
        }
    }
}

pub fn dispatch(action: Action) {
    match action {
        Action::KeyDown(vk) => {
            input::key_down(vk);
        }
        Action::KeyUp(vk) => {
            input::key_up(vk);
        }
        Action::MouseDown(b) => {
            input::mouse_down(b);
        }
        Action::MouseUp(b) => {
            input::mouse_up(b);
        }
    }
}

pub fn describe(action: Action) -> String {
    match action {
        Action::KeyDown(vk) => format!("按下 {}", key_name(vk)),
        Action::KeyUp(vk) => format!("松开 {}", key_name(vk)),
        Action::MouseDown(b) => format!("按下鼠标 {}", button_name(b)),
        Action::MouseUp(b) => format!("松开鼠标 {}", button_name(b)),
    }
}

pub fn button_name(b: MouseButton) -> &'static str {
    match b {
        MouseButton::Left => "左键",
        MouseButton::Middle => "中键",
        MouseButton::Right => "右键",
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::score::parse;

    fn actions_of(score: &str) -> Vec<TimedAction> {
        let song = parse(score).unwrap();
        let cfg = Config::default();
        build_actions(&song, &cfg, 0)
    }

    #[test]
    fn note_produces_down_then_up() {
        let acts = actions_of("bpm=120\nticks_per_beat=12\n[score]\nM1/12");
        let downs: Vec<_> = acts
            .iter()
            .filter(|a| matches!(a.action, Action::KeyDown(_)))
            .collect();
        let ups: Vec<_> = acts
            .iter()
            .filter(|a| matches!(a.action, Action::KeyUp(_)))
            .collect();
        assert_eq!(downs.len(), 1);
        assert_eq!(ups.len(), 1);
        assert!(ups[0].at_ms > downs[0].at_ms);
        // 中音区不需要鼠标修饰
        assert!(!acts
            .iter()
            .any(|a| matches!(a.action, Action::MouseDown(_))));
    }

    #[test]
    fn low_register_adds_mouse_modifier_before_note() {
        let acts = actions_of("bpm=120\nticks_per_beat=12\n[score]\nL1/12");
        let mouse_down = acts
            .iter()
            .find(|a| matches!(a.action, Action::MouseDown(_)))
            .expect("低音区应按下鼠标修饰键");
        let key_down = acts
            .iter()
            .find(|a| matches!(a.action, Action::KeyDown(_)))
            .unwrap();
        assert!(mouse_down.at_ms < key_down.at_ms);
    }

    #[test]
    fn repeated_identical_notes_are_separated() {
        let acts = actions_of("bpm=120\nticks_per_beat=12\n[score]\nM1/12 M1/12");
        let key_events: Vec<_> = acts
            .iter()
            .filter(|a| matches!(a.action, Action::KeyDown(_) | Action::KeyUp(_)))
            .collect();
        // 同音重复必须 松-按-松-按
        assert_eq!(key_events.len(), 4);
        assert!(matches!(key_events[1].action, Action::KeyUp(_)));
        assert!(key_events[1].at_ms < key_events[2].at_ms);
    }

    #[test]
    fn shared_modifier_stays_held_across_segments() {
        // L1 接 L1#：左键两段都要用，应该从头按到尾，只在中间补按中键
        let acts = actions_of("bpm=120\nticks_per_beat=12\n[score]\nL1/12 L1#/12");
        let left: Vec<_> = acts
            .iter()
            .filter(|a| {
                matches!(
                    a.action,
                    Action::MouseDown(MouseButton::Left) | Action::MouseUp(MouseButton::Left)
                )
            })
            .collect();
        assert_eq!(
            left.len(),
            2,
            "左键应当只按下一次、松开一次，实际动作：{left:?}"
        );
        assert!(matches!(left[0].action, Action::MouseDown(_)));
        assert!(matches!(left[1].action, Action::MouseUp(_)));

        let middle: Vec<_> = acts
            .iter()
            .filter(|a| {
                matches!(
                    a.action,
                    Action::MouseDown(MouseButton::Middle) | Action::MouseUp(MouseButton::Middle)
                )
            })
            .collect();
        assert_eq!(middle.len(), 2, "中键应当在第二段才按下");
    }

    #[test]
    fn octave_modifier_is_held_for_whole_group() {
        // 连续 7 个低音：左键要按住整组，而不是每个音点一下
        let acts = actions_of("bpm=120\nticks_per_beat=12\n[score]\nL1/12 L2/12 L3/12 L4/12 L5/12 L6/12 L7/12");
        let down = acts
            .iter()
            .find(|a| matches!(a.action, Action::MouseDown(MouseButton::Left)))
            .unwrap();
        let up = acts
            .iter()
            .find(|a| matches!(a.action, Action::MouseUp(MouseButton::Left)))
            .unwrap();
        // 7 个音，每音 500ms，总共约 3.5 秒
        assert!(
            up.at_ms - down.at_ms > 3000.0,
            "左键只按住了 {} ms，应当是整组长按",
            up.at_ms - down.at_ms
        );
    }

    #[test]
    fn modifier_waits_until_previous_note_ends() {
        // 低音短音紧接高音：左键要松开、右键要按下，
        // 两个动作都不能早于低音的音符键松开 —— 否则音符还在响时音高就变了。
        let song = parse("bpm=120\nticks_per_beat=24\n[score]\nL1/3 H1/24").unwrap();
        let cfg = Config::default();
        let acts = build_actions(&song, &cfg, 0);
        let safety = cfg.timing.note_gap_ms as f64;

        let note_up = acts
            .iter()
            .find(|a| matches!(a.action, Action::KeyUp(_)))
            .expect("低音应当有松开");
        let left_up = acts
            .iter()
            .find(|a| matches!(a.action, Action::MouseUp(MouseButton::Left)))
            .expect("左键应当被松开");
        let right_down = acts
            .iter()
            .find(|a| matches!(a.action, Action::MouseDown(MouseButton::Right)))
            .expect("右键应当被按下");

        assert!(
            left_up.at_ms >= note_up.at_ms + safety,
            "左键在 {:.0}ms 就松开了，而音符键 {:.0}ms 才松开，会发出难听的滑音",
            left_up.at_ms,
            note_up.at_ms
        );
        assert!(
            right_down.at_ms >= note_up.at_ms + safety,
            "右键在 {:.0}ms 就按下了，而音符键 {:.0}ms 才松开",
            right_down.at_ms,
            note_up.at_ms
        );
    }

    #[test]
    fn modifier_release_never_beats_note_release() {
        // 整组低音结束时，修饰键的松开也必须晚于最后一个音的松开
        let song =
            parse("bpm=120\nticks_per_beat=24\n[score]\nL1/3 L2/3 L3/24").unwrap();
        let cfg = Config::default();
        let acts = build_actions(&song, &cfg, 0);

        let last_note_up = acts
            .iter()
            .filter(|a| matches!(a.action, Action::KeyUp(_)))
            .map(|a| a.at_ms)
            .fold(f64::MIN, f64::max);
        let left_up = acts
            .iter()
            .find(|a| matches!(a.action, Action::MouseUp(MouseButton::Left)))
            .expect("左键应当被松开");

        assert!(
            left_up.at_ms >= last_note_up,
            "左键在 {:.0}ms 松开，最后一个音 {:.0}ms 才松",
            left_up.at_ms,
            last_note_up
        );
    }

    #[test]
    fn rest_produces_no_note_actions() {
        let acts = actions_of("bpm=120\nticks_per_beat=12\n[score]\nR/12");
        assert!(!acts
            .iter()
            .any(|a| matches!(a.action, Action::KeyDown(_) | Action::MouseDown(_))));
    }
}

File: core/player.cpp
#include "player.h"

#include <algorithm>
#include <chrono>
#include <thread>

#include "input.h"

namespace harmonica {
namespace {

/// 按键时长的数学下限（毫秒）。时值比 `noteGapMs` 还短的音会被压到这里，
/// 保证「按下」之后一定有对应的「松开」—— 否则键会卡在按下状态。
/// 这不是保底，是防止动作表出现「先松后按」的硬约束。
constexpr double MIN_KEY_HOLD_MS = 1.0;

struct PlayedNote {
    double startMs = 0.0;
    double durMs = 0.0;
    std::uint16_t vk = 0;
    std::vector<MouseButton> buttons;
};

/// 把演奏时刻和一个动作拼成表项：`at<MouseUp>(ms, MouseButton::Left)`
template <typename ActionT, typename ValueT>
TimedAction at(double atMs, ValueT value) {
    return TimedAction{atMs, ActionT{value}};
}

}  // namespace

std::vector<TimedAction> buildActions(const Song& song, const Config& cfg) {
    const double msPerTick = song.msPerTick();
    std::vector<PlayedNote> notes;
    // 时间轴整体后移一个 lead，保证第一个音的鼠标修饰键也能提前按下
    double cursor = static_cast<double>(cfg.timing.mouseLeadMs);

    for (const Event& event : song.events) {
        const double dur = static_cast<double>(ticksOf(event)) * msPerTick;
        if (const auto* note = std::get_if<NoteEvent>(&event)) {
            const auto [vk, buttons] = cfg.resolve(note->pitch);
            notes.push_back(PlayedNote{cursor, dur, vk, buttons});
        }
        cursor += dur;
    }

    const double lead = static_cast<double>(cfg.timing.mouseLeadMs);
    const double gap = static_cast<double>(cfg.timing.noteGapMs);

    std::vector<TimedAction> actions;
    // 修饰键按"差集"维护：本段仍然需要的键保持按住不撒手，
    // 只松开不再需要的、按下新需要的。
    std::vector<MouseButton> held;
    double lastOff = 0.0;
    // 上一段最后一个音符键的松开时刻。这一段修饰键要动，必须等它过去。
    double prevLastKeyUp = 0.0;
    std::size_t i = 0;

    while (i < notes.size()) {
        const std::vector<MouseButton> buttons = notes[i].buttons;
        const auto contains = [&buttons](MouseButton b) {
            return std::find(buttons.begin(), buttons.end(), b) != buttons.end();
        };
        std::size_t j = i;
        while (j + 1 < notes.size() && notes[j + 1].buttons == buttons) ++j;

        // 修饰键最早能动的时刻：前一个音彻底结束，再留 gap 的余量。
        // 早了的话，音符还在响、音高就变了，会发出一个很难听的滑音。
        const double earliest = i > 0 ? prevLastKeyUp + gap : 0.0;
        const double onAt = std::max(notes[i].startMs - lead, earliest);

        // 先松开本段用不到的，再按下本段新增的（同刻时 actionOrder 保证这个先后）。
        // 游戏按事件逐个处理：新的修饰键按下时旧的还没松，它就会同时看到
        // 两组修饰键（比如 左键+右键），这一段的音高就全错了。
        for (const MouseButton b : held) {
            if (!contains(b)) actions.push_back(at<MouseUp>(onAt, b));
        }
        for (const MouseButton b : buttons) {
            const bool alreadyHeld =
                std::find(held.begin(), held.end(), b) != held.end();
            if (!alreadyHeld) actions.push_back(at<MouseDown>(onAt, b));
        }
        held = buttons;

        // 音符键。段首那一个必须等修饰键建立起来（onAt + lead），
        // 否则开头一小段是按在错误的音区上，听起来就是跑调。
        double lastKeyUp = 0.0;
        for (std::size_t index = i; index <= j; ++index) {
            const PlayedNote& note = notes[index];
            // 段内也必须一个接一个：修饰键被推迟时（时值比松键间隔还短的音一定会遇到），
            // 只推迟段首会让后面的音跑到修饰键前面去，也可能早于本段第一个音 ——
            // 那既串了音，又按在错误的音区上。正常谱子里 `前一个音的松开 + gap`
            // 正好等于本音的记谱时刻，这条 max 不会改变任何东西。
            const double earliestDown =
                index == i ? onAt + lead : lastKeyUp + gap;
            const double downAt = std::max(note.startMs, earliestDown);
            // 完全按谱面时值：按住 = 时值 - 松键间隔
            const double upAt = downAt + std::max(note.durMs - gap, MIN_KEY_HOLD_MS);

            actions.push_back(at<KeyDown>(downAt, note.vk));
            actions.push_back(at<KeyUp>(upAt, note.vk));
            lastKeyUp = upAt;
        }
        prevLastKeyUp = lastKeyUp;
        // 修饰键的使命到**最后一个音的键松开**就结束了。
        // 以前这里多留了一个 lead（音结束再等一会儿才松），结果下一段的修饰键
        // 提前 lead 按下时旧的还没松 —— 游戏同时看到两组修饰键，音高判定作废。
        // 键和修饰键同一刻松开没问题：actionOrder 保证先松键、后松修饰键。
        lastOff = lastKeyUp;
        i = j + 1;
    }

    // 收尾：松开还按着的修饰键
    for (const MouseButton b : held) actions.push_back(at<MouseUp>(lastOff, b));

    std::stable_sort(actions.begin(), actions.end(), [](const TimedAction& a, const TimedAction& b) {
        if (a.atMs != b.atMs) return a.atMs < b.atMs;
        return actionOrder(a.action) < actionOrder(b.action);
    });
    return actions;
}

int actionOrder(const Action& action) {
    if (std::holds_alternative<KeyUp>(action)) return 0;
    if (std::holds_alternative<MouseUp>(action)) return 1;
    if (std::holds_alternative<MouseDown>(action)) return 2;
    return 3;  // KeyDown
}

QString describe(const Action& action) {
    return std::visit(
        [](const auto& a) -> QString {
            using T = std::decay_t<decltype(a)>;
            if constexpr (std::is_same_v<T, KeyDown>) {
                return QStringLiteral("key down %1").arg(keyName(a.vk));
            } else if constexpr (std::is_same_v<T, KeyUp>) {
                return QStringLiteral("key up %1").arg(keyName(a.vk));
            } else if constexpr (std::is_same_v<T, MouseDown>) {
                return QStringLiteral("mouse down %1").arg(buttonName(a.button));
            } else {
                return QStringLiteral("mouse up %1").arg(buttonName(a.button));
            }
        },
        action);
}

void dispatch(const Action& action) {
    std::visit(
        [](const auto& a) {
            using T = std::decay_t<decltype(a)>;
            if constexpr (std::is_same_v<T, KeyDown>) {
                input::keyDown(a.vk);
            } else if constexpr (std::is_same_v<T, KeyUp>) {
                input::keyUp(a.vk);
            } else if constexpr (std::is_same_v<T, MouseDown>) {
                input::mouseDown(a.button);
            } else {
                input::mouseUp(a.button);
            }
        },
        action);
}

ReleaseGuard::ReleaseGuard(std::vector<std::uint16_t> keys, bool active)
    : keys_(std::move(keys)), active_(active) {}

ReleaseGuard::~ReleaseGuard() {
    if (active_) input::releaseAll(keys_);
}

void play(const std::vector<TimedAction>& actions, const Config& cfg,
          std::atomic<bool>& stopRequested, bool dryRun,
          std::atomic<std::uint64_t>* progress) {
    // 无论正常结束、提前 return 还是抛异常，离开作用域时都会松开所有按键。
    // 漏掉这一步的后果很严重：鼠标键会卡在按下状态，导致整个系统点不动。
    const std::vector<std::uint16_t> keys = cfg.allKeys();
    const ReleaseGuard guard{keys, !dryRun};

    const auto base = std::chrono::steady_clock::now();
    std::size_t index = 0;

    while (true) {
        if (stopRequested.load(std::memory_order_relaxed)) break;
        if (index >= actions.size()) break;

        const auto target = base + std::chrono::duration_cast<std::chrono::steady_clock::duration>(
                                        std::chrono::duration<double, std::milli>(actions[index].atMs));
        const auto now = std::chrono::steady_clock::now();
        if (now < target) {
            // 最多睡 20ms 就回头看一眼停止标志，睡死了 F10 就按不停
            const auto remainMs =
                std::chrono::duration_cast<std::chrono::milliseconds>(target - now).count();
            std::this_thread::sleep_for(std::chrono::milliseconds(std::min<std::int64_t>(remainMs, 20)));
            continue;
        }

        const TimedAction& timed = actions[index];
        if (!dryRun) dispatch(timed.action);
        if (progress != nullptr) {
            progress->store(static_cast<std::uint64_t>(timed.atMs), std::memory_order_relaxed);
        }
        ++index;
    }
}

const std::vector<CalibrationGroup>& calibrationGroups() {
    // 每个音位都单独走 Config::resolve，不假设整组共用一套修饰键。
    static const std::vector<CalibrationGroup> groups{
        {Register::Mid, false, 7, QStringLiteral("不修饰")},
        {Register::Mid, true, 7, QStringLiteral("半音")},
        {Register::Low, false, 7, QStringLiteral("降调")},
        {Register::Low, true, 7, QStringLiteral("降调 + 半音")},
        {Register::High, false, 7, QStringLiteral("升调")},
        {Register::High, true, 7, QStringLiteral("升调 + 半音")},
        {Register::Top, false, 1, QStringLiteral("升调 + 逗号键")},
        {Register::Top, true, 1, QStringLiteral("升调 + 半音 + 逗号键")},
    };
    return groups;
}

std::vector<CalibrationStep> calibrationSteps(const Config& cfg) {
    std::vector<CalibrationStep> out;
    for (const CalibrationGroup& group : calibrationGroups()) {
        for (std::uint8_t degree = 1; degree <= group.count; ++degree) {
            const Pitch pitch{group.register_, degree, group.sharp};
            const auto [vk, buttons] = cfg.resolve(pitch);
            out.push_back(CalibrationStep{group.name, degree, vk, buttons});
        }
    }
    return out;
}

Song calibrationSong() {
    Song song;
    song.title = QStringLiteral("键位校准");
    song.bpm = 120.0;
    song.ticksPerBeat = 12;
    for (const CalibrationGroup& group : calibrationGroups()) {
        for (std::uint8_t degree = 1; degree <= group.count; ++degree) {
            song.events.push_back(NoteEvent{Pitch{group.register_, degree, group.sharp}, 14});
            song.events.push_back(RestEvent{5});
        }
    }
    return song;
}

}  // namespace harmonica
